rec/utils.py

```diff
@@ -9,13 +9,66 @@
 
 
 def normalize_matrix(matrix, axis=1):
-    divisor = matrix.sum(axis=axis)[:, None]
-    result = np.divide(
-        matrix, divisor, out=np.zeros(matrix.shape, dtype=float), where=divisor != 0
-    )
+    """ Normalize a matrix so that each row vector has a Euclidean norm of 1.
+        If a vector is passed in, we treat it as a matrix with a single row.
+    """
+    if len(matrix.shape) == 1:
+        # turn vector into matrix with one row
+        matrix = matrix[:, np.newaxis]
+    divisor = np.linalg.norm(matrix, axis=1)[:, np.newaxis]
+    # only normalize where divisor is not zero
+    result = np.divide(matrix, divisor, out=np.zeros(matrix.shape), where=divisor != 0)
     return result
 
 
+def contains_row(matrix, row):
+    """ Check if a numpy matrix contains a row with the same values as the 
+        variable `row`.
+    """
+    return (matrix == row).all(1).any()
+
+
+def slerp(mat1, mat2, t=0.05):
+    """ Implements `spherical linear interpolation`_. Takes each row vector in
+        mat1 and rotates it in the direction of the corresponding row vector in
+        mat2. The angle of rotation is `(t * the angle between the two row
+        vectors)`. i.e., when `t=0.05`, each row vector
+        in mat1 is rotated with an angle equal to 5% of the total angle
+        between it and the corresponding row vector in mat2. The matrix returned
+        will have row vectors that each have the same norm as mat1, but pointing
+        in different directions.
+
+        .. _`spherical linear interpolation`: https://en.wikipedia.org/wiki/Slerp
+
+        Parameters
+        -----------
+
+            mat1: numpy.ndarray or list
+                Matrix whose row vectors will be rotated in the direction of
+                mat2's row vectors.
+            
+            mat2: numpy.ndarray or list
+                Matrix that should have the same dimensions at mat1.
+
+            t: float
+                Parameter in [0,1] inclusive that specifies the percentage
+                of rotation.
+    """
+    assert t >= 0.0 and t <= 1.0
+    mat1_norm = np.linalg.norm(mat1, axis=1)[:, np.newaxis]
+    mat2_norm = np.linalg.norm(mat2, axis=0)[:, np.newaxis]
+    # dot every user profile with its corresponding item attributes
+    omega = np.arccos((mat1/mat1_norm) * (mat2/mat2_norm).sum(axis=1))
+    # note: bad things will happen if the vectors are in exactly opposite
```

Comment:
Can you raise exception if this happens?

Reply:
mm yes good point!